File: _site/python/category_theory.py
# Objects and morphisms in Category Theory
class Object:
    def __init__(self, name):
        self.name = name

class Morphism:
    def __init__(self, source, target, func):
        self.source = source
        self.target = target
        self.func = func

# Functor: a mapping between categories
class Functor:
    def __init__(self, fmap_obj, fmap_mor):
        self.fmap_obj = fmap_obj
        self.fmap_mor = fmap_mor

def identity(obj):
    return Morphism(obj, obj, lambda x: x)

def compose(m1, m2):
    return Morphism(m2.source, m1.target, lambda x: m1.func(m2.func(x)))

def test_category_theory():
    obj_a = Object("A")
    obj_b = Object("B")
    morphism = Morphism(obj_a, obj_b, lambda x: obj_b)
    functor = Functor(lambda o: o, lambda m: m)
    print(f"Functor maps object: {functor.fmap_obj(obj_a)}")

File: haskell/CategoryTheory.hs
module CategoryTheory where

-- Defining an object as an abstract type
data Object = Object String deriving (Show, Eq)

-- A morphism is a function between objects
data Morphism = Morphism Object Object (Object -> Object)

-- Custom Show instance for Morphism
instance Show Morphism where
  show (Morphism source target _) = "Morphism from " ++ show source ++ " to " ++ show target

-- Functor: A mapping between categories
data Functor = Functor {
  fmapObj :: Object -> Object,
  fmapMor :: Morphism -> Morphism
}

-- Identity morphism
identity :: Object -> Morphism
identity o = Morphism o o id

-- Composition of morphisms
compose :: Morphism -> Morphism -> Maybe Morphism
compose (Morphism _ _ f) (Morphism o2 o3 g) = Just (Morphism o2 o3 (f . g))

-- Define SpaceTimeFunctor: maps objects and morphisms in spacetime context
data SpaceTimeFunctor = SpaceTimeFunctor {
  spaceFmapObj :: Object -> Object,          -- Maps spacetime objects
  spaceFmapMor :: Morphism -> Morphism       -- Maps spacetime morphisms
}

-- Example of how SpaceTimeFunctor might transform an object and a morphism
applySpaceTimeFunctor :: SpaceTimeFunctor -> Object -> Morphism -> (Object, Morphism)
applySpaceTimeFunctor functor obj mor = (spaceFmapObj functor obj, spaceFmapMor functor mor)

-- Test for Category Theory with SpaceTimeFunctor
testCategoryTheory :: IO ()
testCategoryTheory = do
  let objA = Object "A"
  let objB = Object "B"
  let morphism = Morphism objA objB (\_ -> objB)
  
  -- Define a simple SpaceTimeFunctor that maps each object to itself and each morphism to itself
  let spaceTimeFunctor = SpaceTimeFunctor id (\m -> m)
  
  -- Applying SpaceTimeFunctor to objA and morphism
  let (mappedObj, mappedMorphism) = applySpaceTimeFunctor spaceTimeFunctor objA morphism
  
  putStrLn $ "Mapped Object: " ++ show mappedObj
  putStrLn $ "Mapped Morphism: " ++ show mappedMorphism
